File: sdk/csharp/LibraVDB/Collection.cs
using System;
using System.Text.Json;
using System.Collections.Generic;

namespace LibraVDB
{
    public class Collection
    {
        private readonly int _colID;
        private readonly string _name;
        private readonly int _dimension;

        internal Collection(int colID, string name, int dimension)
        {
            _colID = colID;
            _name = name;
            _dimension = dimension;
        }

        public void Insert(string id, float[] vector, object metadata = null)
        {
            if (vector.Length != _dimension)
            {
                throw new ArgumentException($"Vector dimension mismatch. Expected {_dimension}, got {vector.Length}");
            }

            string metaStr = metadata != null ? JsonSerializer.Serialize(metadata) : "{}";

            IntPtr errPtr = Bindings.InsertVector(_colID, id, vector, _dimension, metaStr);
            Database.CheckError(errPtr, "Failed to insert vector");
        }
        
        public void UpdateIfVersion(string id, float[] vector, ulong expectedVersion, object metadata = null)
        {
            if (vector.Length != _dimension)
            {
                throw new ArgumentException($"Vector dimension mismatch. Expected {_dimension}, got {vector.Length}");
            }

            string metaStr = metadata != null ? JsonSerializer.Serialize(metadata) : "{}";

            IntPtr errPtr = Bindings.UpdateVectorIfVersion(_colID, id, vector, _dimension, metaStr, expectedVersion);
            Database.CheckError(errPtr, "Failed to update vector");
        }

        public unsafe void InsertBatch(string[] ids, float[][] vectors, object[] metadata = null)
        {
            if (ids.Length != vectors.Length || (metadata != null && ids.Length != metadata.Length))
            {
                throw new ArgumentException("Length of ids, vectors, and metadata must match");
            }

            int count = ids.Length;
            if (count == 0) return;

            string[] metas = new string[count];
            float[] flatVectors = new float[count * _dimension];

            for (int i = 0; i < count; i++)
            {
                if (vectors[i].Length != _dimension)
                {
                    throw new ArgumentException($"Vector at index {i} has dimension {vectors[i].Length}, expected {_dimension}");
                }

                metas[i] = (metadata != null && metadata[i] != null) ? JsonSerializer.Serialize(metadata[i]) : "{}";
                Array.Copy(vectors[i], 0, flatVectors, i * _dimension, _dimension);
            }

            fixed (float* pVectors = flatVectors)
            {
                IntPtr errPtr = Bindings.InsertBatch(_colID, ids, pVectors, count, _dimension, metas);
                Database.CheckError(errPtr, "Failed to insert batch");
            }
        }

        public string Search(float[] vector, int k, Filter filter = null)
        {
            if (vector.Length != _dimension)
            {
                throw new ArgumentException($"Vector dimension mismatch. Expected {_dimension}, got {vector.Length}");
            }

            string filterJson = filter != null ? filter.ToJson() : "{}";

            IntPtr resPtr = Bindings.QueryVector(_colID, vector, _dimension, k, filterJson);
            return Database.ExtractString(resPtr);
        }

        public string Scan(int offset = 0, int limit = 100)
        {
            IntPtr resPtr = Bindings.ScanCollection(_colID, offset, limit);
            return Database.ExtractString(resPtr);
        }

        public string Get(string id)
        {
            IntPtr resPtr = Bindings.GetVector(_colID, id);
            return Database.ExtractString(resPtr);
        }

        public void Delete(string id)
        {
            IntPtr errPtr = Bindings.DeleteVector(_colID, id);
            Database.CheckError(errPtr, "Failed to delete vector");
        }

        public void DeleteBatch(string[] ids)
        {
            IntPtr errPtr = Bindings.DeleteBatch(_colID, ids, ids.Length);
            Database.CheckError(errPtr, "Failed to delete batch");
        }
    }
}

File: sdk/swift/Sources/LibraVDB/Collection.swift
import Foundation
import CLibraVDB

public class Collection {
    private let dbID: Int32
    private let colID: Int32
    public let name: String
    public let dimension: Int

    internal init(dbID: Int32, colID: Int32, name: String, dimension: Int) {
        self.dbID = dbID
        self.colID = colID
        self.name = name
        self.dimension = dimension
    }

    public func insert(id: String, vector: [Float], metadata: [String: Any]? = nil) throws {
        guard vector.count == dimension else {
            throw LibraVDBError.nativeError("Vector dimension mismatch. Expected \(dimension), got \(vector.count)")
        }

        let metaStr = metadataToJson(metadata)
        
        try id.withCString { cId in
            try metaStr.withCString { cMeta in
                var mutableVec = vector
                try mutableVec.withUnsafeMutableBufferPointer { vecBuffer in
                    if let ptr = vecBuffer.baseAddress {
                        let errPtr = InsertVector(colID, UnsafeMutablePointer(mutating: cId), ptr, Int32(dimension), UnsafeMutablePointer(mutating: cMeta))
                        try checkError(errPtr)
                    }
                }
            }
        }
    }

    public func insertBatch(ids: [String], vectors: [[Float]], metadata: [[String: Any]?]? = nil) throws {
        let count = ids.count
        guard vectors.count == count else {
            throw LibraVDBError.nativeError("Length of ids and vectors must match")
        }
        if let metas = metadata, metas.count != count {
            throw LibraVDBError.nativeError("Length of ids and metadata must match")
        }

        var flatVectors = [Float]()
        flatVectors.reserveCapacity(count * dimension)
        
        var cIds = [UnsafeMutablePointer<CChar>?]()
        var cMetas = [UnsafeMutablePointer<CChar>?]()
        
        for i in 0..<count {
            guard vectors[i].count == dimension else {
                throw LibraVDBError.nativeError("Vector at index \(i) has dimension \(vectors[i].count), expected \(dimension)")
            }
            flatVectors.append(contentsOf: vectors[i])
            cIds.append(strdup(ids[i]))
            
            let metaStr = metadataToJson(metadata?[i])
            cMetas.append(strdup(metaStr))
        }

        defer {
            for ptr in cIds { free(ptr) }
            for ptr in cMetas { free(ptr) }
        }

        try flatVectors.withUnsafeMutableBufferPointer { vecBuffer in
            if let vecPtr = vecBuffer.baseAddress {
                try cIds.withUnsafeMutableBufferPointer { idsBuffer in
                    try cMetas.withUnsafeMutableBufferPointer { metasBuffer in
                        if let cIdsPtr = idsBuffer.baseAddress, let cMetasPtr = metasBuffer.baseAddress {
                            let errPtr = InsertBatch(colID, cIdsPtr, vecPtr, Int32(count), Int32(dimension), cMetasPtr)
                            try checkError(errPtr)
                        }
                    }
                }
            }
        }
    }

    public func search(vector: [Float], k: Int, filter: Filter? = nil) throws -> String {
        guard vector.count == dimension else {
            throw LibraVDBError.nativeError("Vector dimension mismatch")
        }

        let filterJson = filter?.toJson() ?? "{}"

        return try filterJson.withCString { cFilter in
            var mutableVec = vector
            return try mutableVec.withUnsafeMutableBufferPointer { vecBuffer in
                if let ptr = vecBuffer.baseAddress {
                    let resPtr = QueryVector(colID, ptr, Int32(dimension), Int32(k), UnsafeMutablePointer(mutating: cFilter))
                    return try extractString(resPtr) ?? "{}"
                }
                return "{}"
            }
        }
    }

    public func scan(offset: Int = 0, limit: Int = 100) throws -> String {
        let resPtr = ScanCollection(colID, Int32(offset), Int32(limit))
        return try extractString(resPtr) ?? "{}"
    }

    public func delete(id: String) throws {
        try id.withCString { cId in
            let errPtr = DeleteVector(colID, UnsafeMutablePointer(mutating: cId))
            try checkError(errPtr)
        }
    }

    private func metadataToJson(_ metadata: [String: Any]?) -> String {
        guard let metadata = metadata else { return "{}" }
        guard let data = try? JSONSerialization.data(withJSONObject: metadata, options: []) else { return "{}" }
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}
